tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -131,29 +130,33 @@ def get_config(self):
         return config
 
     @classmethod
-    def create_optimizer_spec(cls, optimizer_instance, layer):
-
-        assert isinstance(
-            optimizer_instance, tf.keras.optimizers.Optimizer
-        ), "Object passed is not an instance of tf.keras.optimizers.Optimizer"
-
-        assert isinstance(layer, tf.keras.layers.Layer) or isinstance(
-            layer, tf.keras.Model
-        ), "Object passed is not an instance of tf.keras.layers.Layer nor tf.keras.Model"
+    def create_optimizer_spec(
+        cls,
+        optimizer: tf.keras.optimizers.Optimizer,
+        layer: Union[
+            tf.keras.Model,
+            tf.keras.Sequential,
+            tf.keras.layers.Layer,
+            List[tf.keras.layers.Layer],
+        ],
+    ):
+        """Creates a serializable optimizer spec.
 
-        if type(layer) == list:
+        The name of each variable is used rather than `var.ref()` to enable serialization and deserialization.
+        """
+        if isinstance(layer, list):
             weights = [var.name for sublayer in layer for var in sublayer.weights]
```

Comment:
Is here trainable_weights?

Reply:
This is discussed in https://github.com/tensorflow/addons/pull/969#discussion_r487591038.

Reply:
I would respect the decision of code owner. Change in design is not part of this PR. /cc @hyang0129

Reply:
If it was interpreted as we have `weights` and we set the same `weights` at semantic level is ok

Reply:
Yep. We can wait the response from the code owner. I see the underlying `fit` only takes trainable variables into account. As you recommend, we can change it to `trainable_weights` after dicussion.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/engine/training.py#L737

Reply:
I would prefer to carry over any features/bugs inherent in the original tensorflow. I believe that other optimizers do not make the distinction between weights and trainable weights. Rather, the weights themselves contain a trainable attribute that the graph seems to understand.

Reply:
Correct. a common use case in custom training loop is

```python
multi_optimizer.apply_gradient(grads_and_vars, model.trainable_variables)
```

in this case, even though we record the name of non-trainable variables, they will not be used. I would also vote for `weights` instead of `trainable_weights`

Reply:
As we have `apply_gradients` here at the end I don't know If for readibility we are more in a case like  https://github.com/tensorflow/tensorflow/blob/ba6d71b874c954bdd4219da118318fe2710515cf/tensorflow/python/keras/integration_test/function_test.py#L61

Reply:
Eg in Adam what is the receiving `var_list` https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/optimizer.py#L400 ?

Reply:
My opinion is that as we are only overriding/wrapping `apply_gradients`  our `grads_and_vars` arg is expected to be passed in the base class as a returned by `compute_gradients()`:
>   grads_and_vars: List of (gradient, variable) pairs as returned by
        `compute_gradients()`.
 
But base class `compute_gradients` has a default behavior when `var_list`  is  Null as you can see at:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/optimizer.py#L432-L434

So it is not mandatory but we expect that by default `var_list` is variables.trainable_variables() as you can see from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/optimizer.py#L495

But as we are using directly our weights in apply_gradient it is like we are not using the default behavior or at least like as we have called `computer_gradients()` with all the weights in `var_list` .

Reply:
> My opinion is that as we are only overriding/wrapping `apply_gradients` our `grads_and_vars` arg is expected to be passed in the base class as returned by `compute_gradients()`:
> 
> > grads_and_vars: List of (gradient, variable) pairs as returned by
> > `compute_gradients()`.
> 
> But base class `compute_gradients` has a default behavior when `var_list` is Null as you can see at:
> https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/optimizer.py#L432-L434
> 
> So it is not mandatory but we expect that by default `var_list` is variables.trainable_variables() as you can see from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/training/optimizer.py#L495
> 
> But as we are using directly our weights in apply_gradient it is like we are not using the default behavior or at least like as we have called `computer_gradients()` with all the weights in `var_list` .

This is V1 optimizer?

Reply:
Yes this is V1. 
In V2 you need to pass a variable list or a callable as it has no default but they suppose as default example:
https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/python/keras/optimizer_v2/optimizer_v2.py#L118-L120

Instead in our `apply_gradients` we are controlling internally the variable list to pass to the real `apply_gradiennts` of optimizers. It seems to me that we are performing a little bit like `fit->minimize` method or not?

Reply:
We are not “controlling the var list” internally but fetching only vars that users pass in.

Reply:
If we look at the argument ok we receive an user  argument but internally we access to our prepared `spec["weights"]`:
` for name in spec["weights"]:`

Reply:
the multi optimizer stores a reference of weight names. 

when optimizing, it uses this reference to allocate grad var pairs to the correct optimizer. 

it does not determine what weights are passed to the multi optimizer. 

So you may have a situation where 5 vars are passed but spec['weights'] contains 10 vars and only 3 of them match. Thus, only 3 are passed on. 

The original design is intended to be used in the model.fit function, not a custom optimization loop. Even in a custom optimization loop, the user is expected to specify the var list to include only trainable weights. I am fairly certain that when an optimizer is called in a fit loop, the var list passed includes only trainable variables. Thus, the var list passed to the multi optimizer includes only trainable variables. Finally, the multi optimizer only passes on to its sub optimizers the variables in var list it received that match a variable in spec['weights'] for that particular optimizer. If this is true, then a non trainable variable will never make it to the multi optimizer or sub optimizers. 

This behavior has been tested in a [colab notebook](https://colab.research.google.com/drive/1NKAsrjj1qcls6p8S5dvJT1PkdC2Ieazc?usp=sharing).

Reply:
If it Is a case where weight is a superset to check the name probably ok.

But I still don't catch  when `if var.name == name` needs to handle not trainable weights.